File: csharp/2022/04.cs
namespace AoCSolutions.Year2022;

using Line = ValueTuple<Range, Range>;

public record struct Range(byte Start, byte Stop);

class Day04 : AoCRunner<IEnumerable<Line>>, IAoCDay
{
    public override (object?, object?) Run(IEnumerable<Line> data)
    {
        var countSubset = 0;
        var countIntersection = 0;

        foreach (var (a, b) in data)
        {
            if (a.Start <= b.Start && b.Stop <= a.Stop || b.Start <= a.Start && a.Stop <= b.Stop)
                countSubset++;

            if (b.Start <= a.Stop && a.Start <= b.Stop || a.Start <= b.Stop && b.Start <= a.Stop)
                countIntersection++;
        }

        return (countSubset, countIntersection);
    }

    public override IEnumerable<Line> Process(IEnumerable<string> data) => data.Select(ProcessLine);

    public Line ProcessLine(string line)
    {
        var tokens = line.Split(',', count: 2);
        return (GetRange(tokens[0]), GetRange(tokens[1]));
    }

    public static Range GetRange(string range)
    {
        var tokens = range.Split('-', count: 2);
        var start = byte.Parse(tokens[0]);
        var stop = byte.Parse(tokens[1]);

        return new Range(start, stop);
    }
}

File: python/2022/04.py
from __future__ import annotations

from collections import namedtuple

Range = namedtuple("Range", ["start", "stop"])


def process_line(line):
    def make_range(elf):
        start, _, stop = elf.partition("-")
        return Range(int(start), int(stop))

    first, _, second = line.partition(",")
    return make_range(first), make_range(second)


def run(data):
    count_a = count_b = 0
    for a, b in data:
        if a.start <= b.start and b.stop <= a.stop or b.start <= a.start and a.stop <= b.stop:
            count_a += 1

        if b.start <= a.stop and a.start <= b.stop or a.start <= b.stop and b.start <= a.stop:
            count_b += 1

    yield count_a
    yield count_b
